refactor: nodata handling #162

Kirill888 opened the pull request:
- adding new types
  - ~`MaybeAutoNodata`~ `SomeNodata` -- `None|int|float|str|"auto"`
  - `Nodata` is now `None|float|int`
  - `MaybeNodata` is now ~`None|float|int|str`~  have been removed
  - "auto" replaces what used to be `None`
  - `None` now means "no nodata value"
- `resolve_nodata()` is used to handle nodata options consistently across the library
- `.odc.nodata` is used to extract `nodata` from xarrays
- fixes in reproject/overview generation for float data. It is assumed to have `nan` values, and GDAL needs `nodata=nan` to handle it correctly.

### nodata types explained

- ~`MaybeAutoNodata`~`SomeNodata` should be used in APIs that allow configuring nodata and nodata overrides
- `Nodata` is a type suitable to pass to GDAL `{src,dst)_nodata` it can be `None` which means expect no missing values in the data.
- `FillValue` is a type suitable for `np.full` and is resolved from `Nodata` + `dtype`
   - `NaN` for not-configured floating point images
   - `0` for not-configured integer images
   - value of `nodata` if set (`NaN` is fine here too) 

```
SomeNodata + dtype + [fallback] -> Nodata + dtype -> FillValue
```

Kirill888 committed:
refactor: nodata handling #162

- adding new types
  - `MaybeAutoNodata` -- `None|int|float|str|"auto"`
  - `Nodata` is now `None|float|int`
  - `MaybeNodata` is now `None|float|int|str
  - "auto" replaces what used to be `None`
  - `None` now means "no nodata value"
- `resolve_nodata()` is used to handle nodata
   options consistently across the library
- default nodata for float is `nan`
- fixes in reproject/overview generation for float
  data. It is assumed to have `nan` values, and
  GDAL needs `nodata=nan` to handle it correctly.

SpacemanPaul commented:
Thanks for the swift action on this Kirill - I will review tomorrow.

Kirill888 committed:
Update GH actions

- bump versions for netlify
- supply token to codecov

Kirill888 committed:
Release candidate v0.4.7rc1

Kirill888 committed:
Expand coverage for resolve_nodata method

robbibt commented:
Thanks heaps for this @Kirill888 - the test cases appear to match the functionality I was hoping for nicely, but I'll do a "user" test today and verify that TIFFs exported using the updated code work as intended (e.g. in ESRI etc).

Kirill888 commented:
@robbibt thanks, BTW pleas use `pip install odc-geo==0.4.7rc1` for your test (that release is not on conda, only pypi)

robbibt commented:
I think this is working perfectly. For example, we have a dataset with an Xarray `nodata: nan` attribute:
``` 
import datacube
import odc.geo.xr
from datacube.utils.cog import write_cog

dc = datacube.Datacube()

query_params = dict(
    x=(142.13223, 142.65461),
    y=(-32.17591, -32.54618),
    time=("2022", "2022"),
)

ds = dc.load(product="ga_ls8cls9c_gm_cyear_3", measurements=["edev"], **query_params)
da = ds.edev.squeeze()
```
![image](https://github.com/opendatacube/odc-geo/assets/17680388/692eab3d-3ed1-4f57-9b8e-2abc50508cfb)


This gets written out with a GeoTIFF nodata flag `nodata=nan` using both `datacube` and `odc-geo` tooling:
```
da.odc.write_cog("nodata_nan_odcgeo.tif")
write_cog(da, "nodata_nan_datacube.tif")
```
![image](https://github.com/opendatacube/odc-geo/assets/17680388/de99ba3c-5141-41c2-a8bf-dd7e72af0271)
![image](https://github.com/opendatacube/odc-geo/assets/17680388/1afaf412-d076-4151-86e4-77a18b3105a4)


However, if we write out data without a Xarray `nodata: nan` attribute, `datacube` doesn't include a GeoTIFF nodata flag, but now `odc-geo` does!
```
del da.attrs["nodata"]

da.odc.write_cog("nodata_missing_odcgeo.tif")
write_cog(da, "nodata_missing_datacube.tif")
```
![image](https://github.com/opendatacube/odc-geo/assets/17680388/1022abd2-d061-43b4-9e48-ad364781a913)
![image](https://github.com/opendatacube/odc-geo/assets/17680388/eafa6872-b51b-4932-84a1-eecd88468f5a)

I can though return a true missing nodata value like this:
```
da.odc.write_cog("nodata_truenone_odcgeo.tif", nodata=None)
```
![image](https://github.com/opendatacube/odc-geo/assets/17680388/95b61f05-c63a-4080-8f25-e5a21cdcbcee)

SpacemanPaul commented on odc/geo/types.py:
```diff
@@ -21,8 +21,10 @@
+Nodata = Union[float, int, None]
+MaybeNodata = Union[float, int, str, None]
+MaybeAutoNodata = Union[float, int, str, None, Literal["auto"]]
 T = TypeVar("T")
```
`MaybeNodata = Nodata | str` seems unexpected as `Maybe` usually means `or None`.  I understand that str might be necessary for cases where `nodata in ('nan', 'NAN', 'NaN')` but I'm not sure why this gets a "Maybe" label?

SpacemanPaul replied:
Oh I see, there's a conversion function in `math.py`

I still think the naming here is confusing but I don't have a good suggestion for a replacement.  Maybe `ApiNodata`?  As in the nodata type the user sees in the API (as opposed the internal nodata value passed to GDAL)?

Kirill888 replied:
that's more of a compatibility issue, before this `Nodata` could not be `None`, but it could be a `str`, so there was a `MaybeNodata` also. I guess we could delete it assuming nobody was using it.

SpacemanPaul reviewed (approved):
I think some of the naming is confusing, but the approach is sound.

robbibt commented:
I do want to do a quick test of the COG overviews stuff, so will approve this as soon as I've finished that.

robbibt commented:
> I think some of the naming is confusing, but the approach is sound.

I also find "MaybeAutoNodata" a bit confusing - took me a while to work out what it meant.

Kirill888 commented:
> > I think some of the naming is confusing, but the approach is sound.
> 
> I also find "MaybeAutoNodata" a bit confusing - took me a while to work out what it meant.

What about `SomeNodata` instead of `MaybeAutoNodata`, @SpacemanPaul @robbibt

Kirill888 committed:
Add `.odc.nodata = ` setter

SpacemanPaul commented:
> > > I think some of the naming is confusing, but the approach is sound.
> > 
> > 
> > I also find "MaybeAutoNodata" a bit confusing - took me a while to work out what it meant.
> 
> What about `SomeNodata` instead of `MaybeAutoNodata`, @SpacemanPaul @robbibt

Yes `SomeNodata` or even `AnyNodata` works for me.

Kirill888 committed:
fix: PR feedback change type names

Kirill888 committed:
fix: remove no longer relevant MaybeNodata type

now that `Nodata` is allowed to be `None`,
'MaybeNodata' is now confusing, removing it.

Let's assume that it never was a part of the
`odc.geo` API.

Kirill888 commented:
Updated,

- Removed no longer relevant `MaybeNodata` type, if anyone was using that type, they were probably also using `Nodata` type too, and that have changed, so...

- use `SomeNodata` instead of `MaybeAutoNodata` name
- add `nodata` setter

robbibt commented:
OK, overview generation is working much better for our DEA Intertidal example: even if `nodata` is not set, we still get a result that follows GDAL's expected functionality:

`datacube-core` with `nodata=nan` Xarray attribute (overviews generated correctly ✔️):
<img src="https://github.com/opendatacube/odc-geo/assets/17680388/209228fa-3c33-4402-96c5-fb47d312e2bd" width="400">

`datacube-core` with no nodata Xarray attribute (overviews generated incorrectly ❌):
<img src="https://github.com/opendatacube/odc-geo/assets/17680388/b0f136e9-f1c3-447b-9152-07c892088cb3" width="400">

`odc-geo` with `nodata=nan` Xarray attribute (overviews generated correctly ✔️):
<img src="https://github.com/opendatacube/odc-geo/assets/17680388/8e5ac783-fe49-4f1c-923d-b2a2f6072e2f" width="400">

`odc-geo` with no nodata Xarray attribute (overviews generated correctly ✔️):
<img src="https://github.com/opendatacube/odc-geo/assets/17680388/5bc0f4b5-08a3-4168-bbd9-beba50ca607d" width="400">
